source/MaterialXGenShader/GenContext.h

```diff
@@ -1,86 +1,192 @@
-
-/// Class representing an shader generation context for a node.
-///
-/// For some shader generators a node might need customization to it's implementation 
-/// depending on in which context the node is used. This class handles customizations
-/// by adding a suffix to the function name to distinguish between the functions for 
-/// different contexts. I also supports adding extra arguments to the node's function.
-///
-/// An example of where this is required if for BSDF and EDF nodes for HW targets 
-/// where extra arguments are needed to give directions vectors for evaluation. 
-/// For BSDF nodes another use-case is to distinguish between evaluation in a direct lighting
-/// context and an indirect lighting context where different versions of the node's function
-/// is required.
-/// 
-class GenContext
-{
-  public:
-    /// Constructor, set the identifier for this context.
-    GenContext(int id) : _id(id) {}
-
-    /// Return the identifier for this context.
-    int id() const { return _id; }
+class GenContext;
+class GenUserData;
 
-    /// Add an extra argument to be used for the node function in this context.
-    void addArgument(const Argument& arg) { _arguments.push_back(arg); }
+using GenContextPtr = std::shared_ptr<GenContext>;
+using GenUserDataPtr = std::shared_ptr<GenUserData>;
+using ConstGenUserDataPtr = std::shared_ptr<const GenUserData>;
 
-    /// Return a list of extra argument to be used for the node function in this context.
-    const Arguments& getArguments() const { return _arguments; }
-
-    /// Set a function name suffix to be used for the node function in this context.
-    void setFunctionSuffix(const string& suffix) { _functionSuffix = suffix; }
-
-    /// Return the function name suffix to be used for the node function in this context.
-    const string& getFunctionSuffix() const { return _functionSuffix; }
+/// Base class for custom user data needed during shader generation.
+class GenUserData : public std::enable_shared_from_this<GenUserData>
+{
+  public:
+    virtual ~GenUserData() {}
+    
+    /// Return a shared pointer for this object.
+    GenUserDataPtr getSelf()
+    {
+        return shared_from_this();
+    }
+
+    /// Return a shared pointer for this object.
+    ConstGenUserDataPtr getSelf() const
+    {
+        return shared_from_this();
+    }
+
+    /// Return this object cast to a templated type.
+    template<class T> shared_ptr<T> asA()
+    {
+        return std::dynamic_pointer_cast<T>(getSelf());
+    }
+
+    /// Return this object cast to a templated type.
+    template<class T> shared_ptr<const T> asA() const
+    {
+        return std::dynamic_pointer_cast<const T>(getSelf());
+    }
+
+  protected:
+    GenUserData() {}
+};
 
-    /// Add an input suffix to be used for the node function in this context.
+/// A context class for shader generation.
+/// Used for thread local storage of data needed during shader generation.
+class GenContext
```

Comment:
The context class passed around on most shader gen API calls. Holds the shader generator, the options, the node implementation cache, etc., as well as any custom user data.